File: app/src/main/java/com/example/imake/imake/Activity/DAO/ProfActivity.java
package com.example.imake.imake.Activity.DAO;


import androidx.annotation.NonNull;
import java.text.DecimalFormat;
import androidx.appcompat.app.ActionBarDrawerToggle;
import androidx.appcompat.app.AlertDialog;
import androidx.appcompat.app.AppCompatActivity;
import androidx.appcompat.widget.Toolbar;
import androidx.core.app.ActivityCompat;
import androidx.core.content.ContextCompat;
import androidx.drawerlayout.widget.DrawerLayout;

import android.Manifest;
import android.content.ContentResolver;
import android.content.DialogInterface;
import android.content.Intent;
import android.content.SharedPreferences;
import android.content.pm.PackageManager;
import android.database.Cursor;
import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.net.Uri;
import android.os.Bundle;
import android.provider.MediaStore;
import android.util.Log;
import android.view.MenuItem;
import android.view.View;
import android.webkit.MimeTypeMap;
import android.widget.Button;
import android.widget.ImageButton;
import android.widget.ImageView;
import android.widget.TextView;
import android.widget.Toast;

import com.bumptech.glide.Glide;
import com.example.imake.R;
import com.example.imake.imake.Activity.Activity.FotosProfActivity;
import com.example.imake.imake.Activity.Activity.OpcoesContaActivity;
import com.example.imake.imake.Activity.Entidades.Usuario;
import com.google.android.gms.tasks.OnCompleteListener;
import com.google.android.gms.tasks.OnFailureListener;
import com.google.android.gms.tasks.OnSuccessListener;
import com.google.android.gms.tasks.Task;
import com.google.android.material.navigation.NavigationView;
import com.google.firebase.FirebaseApp;
import com.google.firebase.auth.AuthCredential;
import com.google.firebase.auth.EmailAuthProvider;
import com.google.firebase.auth.FirebaseAuth;
import com.google.firebase.auth.FirebaseUser;
import com.google.firebase.database.DataSnapshot;
import com.google.firebase.database.DatabaseError;
import com.google.firebase.database.DatabaseReference;
import com.google.firebase.database.FirebaseDatabase;
import com.google.firebase.database.Query;
import com.google.firebase.database.ValueEventListener;
import com.google.firebase.storage.FirebaseStorage;
import com.google.firebase.storage.StorageReference;
import com.google.firebase.storage.UploadTask;

import de.hdodenhof.circleimageview.CircleImageView;

public class ProfActivity extends AppCompatActivity {
    private final int GALERIA_IMAGENS = 1;
    private final int PERMISSAO_REQUEST = 2;

    private FirebaseDatabase bd;
    private DatabaseReference referenceBd;
    private DatabaseReference referenceBd2;
    private StorageReference mStorageRef;
    private DatabaseReference mDatabaseRef;
    Uri selectedImage;

    Usuario u;

    String codigoSessao;

    TextView txtNome;
    TextView txtEmail;
    TextView txtNMakes;
    TextView txtCategoria;
    TextView txtVerLocalizacao;
    Button btnVerVideos;
    TextView txtAlterarFotos;
    Button btnVerFotos;
    Button btnConfirmar;
    CircleImageView imgPerfil;
    ImageButton imgbConfirmar;
    Button btnOpConta;
    ImageView imgStar1;
    ImageView imgStar2;
    ImageView imgStar3;
    ImageView imgStar4;
    ImageView imgStar5;
    TextView txtAvaliacao;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_prof);

        getSupportActionBar().hide();

        if (ContextCompat.checkSelfPermission(this,
                Manifest.permission.READ_EXTERNAL_STORAGE)
                != PackageManager.PERMISSION_GRANTED) {
            if (ActivityCompat.shouldShowRequestPermissionRationale(this,
                    Manifest.permission.READ_EXTERNAL_STORAGE)) {
            } else {
                ActivityCompat.requestPermissions(this,
                        new String[]{Manifest.permission.READ_EXTERNAL_STORAGE},
                        PERMISSAO_REQUEST);
            }
        }
        inicializarFirebase();
        inicializaComponentes();

        Intent intent = this.getIntent();
        Bundle bundle = intent.getExtras();
        codigoSessao = (String) bundle.getSerializable("codigoUsuario");
        bundle.clear();

        pesquisarInfo();


        btnVerFotos.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                Bundle bundle = new Bundle();
                bundle.putString("codigoUsuario", u.getCodigo());
                Intent intent = new Intent(ProfActivity.this, FotosProfActivity.class);
                intent.putExtras(bundle);
                startActivity(intent);
            }
        });
        txtAlterarFotos.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                Intent intent = new Intent(Intent.ACTION_PICK, MediaStore.Images.Media.EXTERNAL_CONTENT_URI);
                startActivityForResult(intent, 1);

            }
        });
        imgbConfirmar.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                uploadFile();
            }
        });
        btnOpConta.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                Bundle bundle = new Bundle();
                bundle.putString("codigoUsuario", u.getCodigo());
                Intent intent = new Intent(ProfActivity.this, OpcoesContaActivity.class);
                intent.putExtras(bundle);
                startActivity(intent);
            }
        });

/*
        btnDeletarConta.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                AlertDialog.Builder builder1 = new AlertDialog.Builder(ProfActivity.this);
                builder1.setMessage("Tem certeza?");
                builder1.setCancelable(true);
                builder1.setPositiveButton(
                        "Sim",
                        new DialogInterface.OnClickListener() {
                            public void onClick(DialogInterface dialog, int id) {
                                final FirebaseUser user = FirebaseAuth.getInstance().getCurrentUser();
                                AuthCredential credential = EmailAuthProvider
                                        .getCredential(codigoSessao, u.getSenha());
                                user.reauthenticate(credential)
                                        .addOnCompleteListener(new OnCompleteListener<Void>() {
                                            @Override
                                            public void onComplete(@NonNull Task<Void> task) {
                                                user.delete()
                                                        .addOnCompleteListener(new OnCompleteListener<Void>() {
                                                            @Override
                                                            public void onComplete(@NonNull Task<Void> task) {
                                                                Log.d("Resultado", "Deletou");
                                                                referenceBd.child("Usuario").child(codigoSessao).removeValue();
                                                                referenceBd.child("Desejos").child(codigoSessao).removeValue();
                                                                referenceBd.child("Favoritos").child(codigoSessao).removeValue();
                                                                referenceBd.child("FotosUsuario").child(codigoSessao).removeValue();
                                                                Query q = referenceBd.child("FotosGeral").orderByChild("codigo").equalTo(codigoSessao);
                                                                q.addValueEventListener(new ValueEventListener() {
                                                                    @Override
                                                                    public void onDataChange(@NonNull DataSnapshot dataSnapshot) {
                                                                        referenceBd.child("Usuario").child(codigoSessao).removeValue();
                                                                        referenceBd.child("Desejos").child(codigoSessao).removeValue();
                                                                        referenceBd.child("Favoritos").child(codigoSessao).removeValue();
                                                                        referenceBd.child("FotosUsuario").child(codigoSessao).removeValue();
                                                                        Query q = referenceBd.child("FotosGeral").orderByChild("codigo").equalTo(codigoSessao);
                                                                        q.addValueEventListener(new ValueEventListener() {
                                                                            @Override
                                                                            public void onDataChange(@NonNull DataSnapshot dataSnapshot) {
                                                                                referenceBd.removeEventListener(this);
                                                                                Intent intent = new Intent(ProfActivity.this, SplashActivity.class);
                                                                                startActivity(intent);
                                                                                SharedPreferences.Editor editor = sessao.edit();
                                                                                editor.remove("codigo");
                                                                                editor.commit();
                                                                                finish();
                                                                            }
                                                                            @Override
                                                                            public void onCancelled(@NonNull DatabaseError databaseError) {
                                                                            }
                                                                        });
                                                                    }
                                                                    @Override
                                                                    public void onCancelled(@NonNull DatabaseError databaseError) {
                                                                    }
                                                                });
                                                                finish();
                                                                Intent intent = new Intent(ProfActivity.this, SplashActivity.class);
                                                                startActivity(intent);
                                                            }
                                                        });
                                            }
                                        });
                            }
                        });
                builder1.setNegativeButton(
                        "Não",
                        new DialogInterface.OnClickListener() {
                            public void onClick(DialogInterface dialog, int id) {
                                dialog.cancel();
                            }
                        });
                AlertDialog alert11 = builder1.create();
                alert11.show();
            }
        });*/
    }

    private String getFileExtension(Uri uri) {
        ContentResolver cR = getContentResolver();
        MimeTypeMap mime = MimeTypeMap.getSingleton();
        return mime.getExtensionFromMimeType(cR.getType(uri));
    }

    private void uploadFile() {
        if (selectedImage != null) {
            final StorageReference fileReference = mStorageRef.child(System.currentTimeMillis()
                    + "." + getFileExtension(selectedImage));

            fileReference.putFile(selectedImage)
                    .addOnSuccessListener(new OnSuccessListener<UploadTask.TaskSnapshot>() {
                        @Override
                        public void onSuccess(UploadTask.TaskSnapshot taskSnapshot) {
                            fileReference.getDownloadUrl().addOnSuccessListener(new OnSuccessListener<Uri>() {
                                @Override
                                public void onSuccess(Uri uri) {
                                    Toast.makeText(ProfActivity.this, "Upload Sucessful", Toast.LENGTH_SHORT).show();
                                    Uri download = uri;
                                    u.setUrlPerfil(download.toString());
                                    mDatabaseRef.child(codigoSessao).setValue(u);
                                }
                            });


                        }
                    })
                    .addOnFailureListener(new OnFailureListener() {
                        @Override
                        public void onFailure(@NonNull Exception e) {
                            Toast.makeText(ProfActivity.this, e.getMessage(), Toast.LENGTH_SHORT).show();
                        }
                    });
        } else {
            Toast.makeText(ProfActivity.this, "No File Selected.", Toast.LENGTH_SHORT).show();
        }
    }

    @Override
    protected void onActivityResult(int requestCode, int resultCode, Intent data) {
        super.onActivityResult(requestCode, resultCode, data);
        if (resultCode == RESULT_OK && requestCode == GALERIA_IMAGENS) {
            selectedImage = data.getData();
            String[] filePath = {MediaStore.Images.Media.DATA};
            Cursor c = getContentResolver().query(selectedImage, filePath, null, null, null);
            c.moveToFirst();
            int columnIndex = c.getColumnIndex(filePath[0]);
            String picturePath = c.getString(columnIndex);
            c.close();
            Bitmap thumbnail = (BitmapFactory.decodeFile(picturePath));
            imgPerfil.setImageBitmap(thumbnail);
        }
    }

    @Override
    public void onRequestPermissionsResult(int requestCode, String permissions[], int[] grantResults) {
        if (requestCode == PERMISSAO_REQUEST) {
            if (grantResults.length > 0
                    && grantResults[0] == PackageManager.PERMISSION_GRANTED) {
                // A permissão foi concedida. Pode continuar
            } else {
                // A permissão foi negada. Precisa ver o que deve ser desabilitado
            }
            return;
        }
    }

    public void inicializaComponentes() {
        txtAvaliacao = (TextView) findViewById(R.id.txtAvaliacao);
        txtNome = (TextView) findViewById(R.id.txtNome);
        txtEmail = (TextView) findViewById(R.id.txtEmail);
        txtCategoria = (TextView) findViewById(R.id.txtCategoria);
        txtNMakes = (TextView) findViewById(R.id.txtNMakes);
        txtAlterarFotos = (TextView) findViewById(R.id.txtAlterarFoto);
        btnVerFotos = (Button) findViewById(R.id.btnVerFotos);
        btnConfirmar = (Button) findViewById(R.id.btnConfirmar);
        imgbConfirmar = (ImageButton) findViewById(R.id.imgbConfirmar);
        imgPerfil = (CircleImageView) findViewById(R.id.imgPerfil);
        btnOpConta = (Button) findViewById(R.id.btnOpConta);

        imgStar1 = (ImageView) findViewById(R.id.imgStar1);
        imgStar2 = (ImageView) findViewById(R.id.imgStar2);
        imgStar3 = (ImageView) findViewById(R.id.imgStar3);
        imgStar4 = (ImageView) findViewById(R.id.imgStar4);
        imgStar5 = (ImageView) findViewById(R.id.imgStar5);
    }

    public void pesquisarImagemPerfil() {
        Query q = referenceBd2.child("Usuario").child(codigoSessao);
        q.addValueEventListener(new ValueEventListener() {
            @Override
            public void onDataChange(@NonNull DataSnapshot dataSnapshot) {
                Usuario u = dataSnapshot.getValue(Usuario.class);
                //String url = String.valueOf(perfil.getImageUrl());
                //String url = dataSnapshot.getValue().toString();
                //url = "https://firebasestorage.googleapis.com/v0/b/imake-b81bb.appspot.com/o/uploads%2F1569829996574.jpg?alt=media&token=05c3bdaa-6728-42d4-845a-b8b5fc63cc43";
                //String url = dataSnapshot.getValue().toString();
                String url = String.valueOf(u.getUrlPerfil());
                Glide.with(getApplicationContext()).load(url).into(imgPerfil);



            }

            @Override
            public void onCancelled(@NonNull DatabaseError databaseError) {

            }
        });
    }

    public void pesquisarInfo() {
        Query q = referenceBd.child("Usuario").child(codigoSessao);
        q.addValueEventListener(new ValueEventListener() {
            @Override
            public void onDataChange(@NonNull DataSnapshot dataSnapshot) {
                u = dataSnapshot.getValue(Usuario.class);
                if(u==null){
                    u = new Usuario();
                }
                txtNome.setText(String.valueOf(u.getNome()));
                txtEmail.setText(String.valueOf(u.getEmail()));
                txtCategoria.setText("Categoria: " + String.valueOf(u.getCategoria()));
                txtNMakes.setText(String.valueOf(u.getNumeroMakes()) + " makes");
                DecimalFormat df = new DecimalFormat("#,###.0");

                txtAvaliacao.setText(String.valueOf(df.format(Double.valueOf(u.getAvaliacao()))));
                if (u.getUrlPerfil() != null) {
                    Glide.with(getApplicationContext()).load(u.getUrlPerfil()).into(imgPerfil);
                }
                if(Double.valueOf(u.getAvaliacao()) == 0){
                    imgStar1.setBackgroundResource(R.drawable.starvoid);
                    imgStar2.setBackgroundResource(R.drawable.starvoid);
                    imgStar3.setBackgroundResource(R.drawable.starvoid);
                    imgStar4.setBackgroundResource(R.drawable.starvoid);
                    imgStar5.setBackgroundResource(R.drawable.starvoid);
                }
                else if(Double.valueOf(u.getAvaliacao())<2 && Double.valueOf(u.getAvaliacao())>=1){
                    imgStar1.setBackgroundResource(R.drawable.staryellow);
                }
                else if(Double.valueOf(u.getAvaliacao())<3 && Double.valueOf(u.getAvaliacao())>=2){
                    imgStar1.setBackgroundResource(R.drawable.staryellow);
                    imgStar2.setBackgroundResource(R.drawable.staryellow);
                }
                else if(Double.valueOf(u.getAvaliacao())<4 && Double.valueOf(u.getAvaliacao())>=3){
                    imgStar1.setBackgroundResource(R.drawable.staryellow);
                    imgStar2.setBackgroundResource(R.drawable.staryellow);
                    imgStar3.setBackgroundResource(R.drawable.staryellow);
                }
                else if(Double.valueOf(u.getAvaliacao())<5 && Double.valueOf(u.getAvaliacao())>=4){
                    imgStar1.setBackgroundResource(R.drawable.staryellow);
                    imgStar2.setBackgroundResource(R.drawable.staryellow);
                    imgStar3.setBackgroundResource(R.drawable.staryellow);
                    imgStar4.setBackgroundResource(R.drawable.staryellow);
                }
                else{
                    imgStar1.setBackgroundResource(R.drawable.staryellow);
                    imgStar2.setBackgroundResource(R.drawable.staryellow);
                    imgStar3.setBackgroundResource(R.drawable.staryellow);
                    imgStar4.setBackgroundResource(R.drawable.staryellow);
                    imgStar5.setBackgroundResource(R.drawable.staryellow);
                }
            }

            @Override
            public void onCancelled(@NonNull DatabaseError databaseError) {

            }
        });
    }

    private void inicializarFirebase() {
        FirebaseApp.initializeApp(ProfActivity.this);
        bd = FirebaseDatabase.getInstance("https://imake-b81bb.firebaseio.com/");
        referenceBd = bd.getReference();
        referenceBd2 = bd.getReference();

        mStorageRef = FirebaseStorage.getInstance().getReference("uploads");
        mDatabaseRef = FirebaseDatabase.getInstance().getReference("Usuario");
    }
}

File: app/src/main/java/com/example/imake/imake/Activity/Activity/ProfActivity.kt
package com.example.imake.imake.Activity.Activity

import java.text.DecimalFormat
import androidx.appcompat.app.AppCompatActivity
import androidx.core.app.ActivityCompat
import androidx.core.content.ContextCompat

import android.Manifest
import android.content.Intent
import android.content.pm.PackageManager
import android.graphics.BitmapFactory
import android.net.Uri
import android.os.Bundle
import android.provider.MediaStore
import android.webkit.MimeTypeMap
import android.widget.Button
import android.widget.ImageButton
import android.widget.ImageView
import android.widget.TextView
import android.widget.Toast

import com.bumptech.glide.Glide
import com.example.imake.R
import com.example.imake.imake.Activity.Entidades.Usuario
import com.google.firebase.FirebaseApp
import com.google.firebase.database.DataSnapshot
import com.google.firebase.database.DatabaseError
import com.google.firebase.database.DatabaseReference
import com.google.firebase.database.FirebaseDatabase
import com.google.firebase.database.ValueEventListener
import com.google.firebase.storage.FirebaseStorage
import com.google.firebase.storage.StorageReference

import de.hdodenhof.circleimageview.CircleImageView

class ProfActivity : AppCompatActivity() {
    private val GALERIA_IMAGENS = 1
    private val PERMISSAO_REQUEST = 2

    private var bd: FirebaseDatabase? = null
    private var referenceBd: DatabaseReference? = null
    private var referenceBd2: DatabaseReference? = null
    private var mStorageRef: StorageReference? = null
    private var mDatabaseRef: DatabaseReference? = null
    private var selectedImage: Uri? = null

    private var u: Usuario? = null

    private lateinit var codigoSessao: String

    private lateinit var txtNome: TextView
    private lateinit var txtEmail: TextView
    private lateinit var txtNMakes: TextView
    private lateinit var txtCategoria: TextView
    private lateinit var txtVerLocalizacao: TextView
    private  var btnVerVideos: Button? = null
    private lateinit var txtAlterarFotos: TextView
    private lateinit var btnVerFotos: Button
    private lateinit var imgPerfil: CircleImageView
    private lateinit var imgbConfirmar: ImageButton
    private lateinit var btnOpConta: Button
    private lateinit var imgStar1: ImageView
    private lateinit var imgStar2: ImageView
    private lateinit var imgStar3: ImageView
    private lateinit var imgStar4: ImageView
    private lateinit var imgStar5: ImageView
    private lateinit var txtAvaliacao: TextView

    private lateinit var intm: Intent

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_prof)

        supportActionBar!!.hide()

        if (ContextCompat.checkSelfPermission(this,
                        Manifest.permission.READ_EXTERNAL_STORAGE) != PackageManager.PERMISSION_GRANTED) {
            if (ActivityCompat.shouldShowRequestPermissionRationale(this,
                            Manifest.permission.READ_EXTERNAL_STORAGE)) {
            } else {
                ActivityCompat.requestPermissions(this,
                        arrayOf(Manifest.permission.READ_EXTERNAL_STORAGE),
                        PERMISSAO_REQUEST)
            }
        }
        inicializarFirebase()
        inicializaComponentes()

        val bundle = intent.extras
        codigoSessao = bundle!!.getSerializable("codigoUsuario") as String
        bundle.clear()

        pesquisarInfo()


        btnVerFotos.setOnClickListener {
            intm = Intent(this@ProfActivity, FotosProfActivity::class.java)
            intm.putExtra("codigoUsuario", u!!.codigo)
            startActivity(intm)
        }
        txtAlterarFotos.setOnClickListener {
            intm = Intent(Intent.ACTION_PICK, MediaStore.Images.Media.EXTERNAL_CONTENT_URI)
            startActivityForResult(intm, 1)
        }
        imgbConfirmar.setOnClickListener { uploadFile() }
        btnOpConta.setOnClickListener {
            intm = Intent(this@ProfActivity, OpcoesContaActivity::class.java)
            intm.putExtra("codigoUsuario", u!!.codigo)
            startActivity(intm)
        }

        /*
           btnDeletarConta.setOnClickListener(new View.OnClickListener() {
               @Override
               public void onClick(View view) {
                   AlertDialog.Builder builder1 = new AlertDialog.Builder(ProfActivity.this);
                   builder1.setMessage("Tem certeza?");
                   builder1.setCancelable(true);

                   builder1.setPositiveButton(
                           "Sim",
                           new DialogInterface.OnClickListener() {
                               public void onClick(DialogInterface dialog, int id) {
                                   final FirebaseUser user = FirebaseAuth.getInstance().getCurrentUser();
                                   AuthCredential credential = EmailAuthProvider
                                           .getCredential(codigoSessao, u.getSenha());

                                   user.reauthenticate(credential)
                                           .addOnCompleteListener(new OnCompleteListener<Void>() {
                                               @Override
                                               public void onComplete(@NonNull Task<Void> task) {
                                                   user.delete()
                                                           .addOnCompleteListener(new OnCompleteListener<Void>() {
                                                               @Override
                                                               public void onComplete(@NonNull Task<Void> task) {
                                                                   Log.d("Resultado", "Deletou");
                                                                   referenceBd.child("Usuario").child(codigoSessao).removeValue();
                                                                   referenceBd.child("Desejos").child(codigoSessao).removeValue();
                                                                   referenceBd.child("Favoritos").child(codigoSessao).removeValue();
                                                                   referenceBd.child("FotosUsuario").child(codigoSessao).removeValue();
                                                                   Query q = referenceBd.child("FotosGeral").orderByChild("codigo").equalTo(codigoSessao);
                                                                   q.addValueEventListener(new ValueEventListener() {
                                                                       @Override
                                                                       public void onDataChange(@NonNull DataSnapshot dataSnapshot) {
                                                                           referenceBd.child("Usuario").child(codigoSessao).removeValue();
                                                                           referenceBd.child("Desejos").child(codigoSessao).removeValue();
                                                                           referenceBd.child("Favoritos").child(codigoSessao).removeValue();
                                                                           referenceBd.child("FotosUsuario").child(codigoSessao).removeValue();
                                                                           Query q = referenceBd.child("FotosGeral").orderByChild("codigo").equalTo(codigoSessao);
                                                                           q.addValueEventListener(new ValueEventListener() {
                                                                               @Override
                                                                               public void onDataChange(@NonNull DataSnapshot dataSnapshot) {
                                                                                   referenceBd.removeEventListener(this);
                                                                                   Intent intent = new Intent(ProfActivity.this, SplashActivity.class);
                                                                                   startActivity(intent);

                                                                                   SharedPreferences.Editor editor = sessao.edit();
                                                                                   editor.remove("codigo");
                                                                                   editor.commit();

                                                                                   finish();
                                                                               }

                                                                               @Override
                                                                               public void onCancelled(@NonNull DatabaseError databaseError) {

                                                                               }
                                                                           });
                                                                       }

                                                                       @Override
                                                                       public void onCancelled(@NonNull DatabaseError databaseError) {

                                                                       }
                                                                   });
                                                                   finish();
                                                                   Intent intent = new Intent(ProfActivity.this, SplashActivity.class);
                                                                   startActivity(intent);


                                                               }
                                                           });
                                               }
                                           });
                               }
                           });

                   builder1.setNegativeButton(
                           "Não",
                           new DialogInterface.OnClickListener() {
                               public void onClick(DialogInterface dialog, int id) {
                                   dialog.cancel();
                               }
                           });

                   AlertDialog alert11 = builder1.create();
                   alert11.show();
               }
           });*/
    }

    private fun getFileExtension(uri: Uri): String? {
        val cR = contentResolver
        val mime = MimeTypeMap.getSingleton()
        return mime.getExtensionFromMimeType(cR.getType(uri))
    }

    private fun uploadFile() {
        if (selectedImage != null) {
            val fileReference = mStorageRef!!.child(System.currentTimeMillis().toString()
                    + "." + getFileExtension(selectedImage!!))

            fileReference.putFile(selectedImage!!)
                    .addOnSuccessListener {
                        fileReference.downloadUrl.addOnSuccessListener { uri ->
                            Toast.makeText(this@ProfActivity, "Upload Sucessful", Toast.LENGTH_SHORT).show()
                            u!!.urlPerfil = uri.toString()
                            mDatabaseRef!!.child(codigoSessao).setValue(u)
                        }
                    }
                    .addOnFailureListener { e -> Toast.makeText(this@ProfActivity, e.message, Toast.LENGTH_SHORT).show() }
        } else {
            Toast.makeText(this@ProfActivity, "No File Selected.", Toast.LENGTH_SHORT).show()
        }
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        super.onActivityResult(requestCode, resultCode, data)
        if (resultCode == RESULT_OK && requestCode == GALERIA_IMAGENS) {
            selectedImage = data!!.data
            val filePath = arrayOf(MediaStore.Images.Media.DATA)
            val c = contentResolver.query(selectedImage!!, filePath, null, null, null)
            c!!.moveToFirst()
            val columnIndex = c.getColumnIndex(filePath[0])
            val picturePath = c.getString(columnIndex)
            c.close()
            val thumbnail = BitmapFactory.decodeFile(picturePath)
            imgPerfil.setImageBitmap(thumbnail)
        }
    }

    override fun onRequestPermissionsResult(requestCode: Int, permissions: Array<String>, grantResults: IntArray) {
        if (requestCode == PERMISSAO_REQUEST) {
            if (grantResults.size > 0 && grantResults[0] == PackageManager.PERMISSION_GRANTED) {
                // A permissão foi concedida. Pode continuar
            } else {
                // A permissão foi negada. Precisa ver o que deve ser desabilitado
            }
            return
        }
    }

    fun inicializaComponentes() {
        txtAvaliacao = findViewById(R.id.txtAvaliacao) as TextView
        txtNome = findViewById(R.id.txtNome) as TextView
        txtEmail = findViewById(R.id.txtEmail) as TextView
        txtCategoria = findViewById(R.id.txtCategoria) as TextView
        txtNMakes = findViewById(R.id.txtNMakes) as TextView
        txtAlterarFotos = findViewById(R.id.txtAlterarFoto) as TextView
        btnVerFotos = findViewById(R.id.btnVerFotos) as Button
        imgbConfirmar = findViewById(R.id.imgbConfirmar) as ImageButton
        imgPerfil = findViewById(R.id.imgPerfil) as CircleImageView
        btnOpConta = findViewById(R.id.btnOpConta) as Button

        imgStar1 = findViewById(R.id.imgStar1) as ImageView
        imgStar2 = findViewById(R.id.imgStar2) as ImageView
        imgStar3 = findViewById(R.id.imgStar3) as ImageView
        imgStar4 = findViewById(R.id.imgStar4) as ImageView
        imgStar5 = findViewById(R.id.imgStar5) as ImageView
    }

    fun pesquisarImagemPerfil() {
        val q = referenceBd2!!.child("Usuario").child(codigoSessao)
        q.addValueEventListener(object : ValueEventListener {
            override fun onDataChange(dataSnapshot: DataSnapshot) {
                val u = dataSnapshot.getValue(Usuario::class.java)
                //String url = String.valueOf(perfil.getImageUrl());
                //String url = dataSnapshot.getValue().toString();
                //url = "https://firebasestorage.googleapis.com/v0/b/imake-b81bb.appspot.com/o/uploads%2F1569829996574.jpg?alt=media&token=05c3bdaa-6728-42d4-845a-b8b5fc63cc43";
                //String url = dataSnapshot.getValue().toString();
                val url = u!!.urlPerfil.toString()
                Glide.with(applicationContext).load(url).into(imgPerfil)


            }

            override fun onCancelled(databaseError: DatabaseError) {

            }
        })
    }

    fun pesquisarInfo() {
        val q = referenceBd!!.child("Usuario").child(codigoSessao)
        q.addValueEventListener(object : ValueEventListener {
            override fun onDataChange(dataSnapshot: DataSnapshot) {
                u = dataSnapshot.getValue(Usuario::class.java)
                if (u == null) {
                    u = Usuario()
                }
                txtNome.text = u!!.nome.toString()
                txtEmail.text = u!!.email.toString()
                txtCategoria.text = "Categoria: " + u!!.categoria.toString()
                txtNMakes.text = u!!.numeroMakes.toString() + " makes"
                val df = DecimalFormat("#,###.0")

                txtAvaliacao.text = df.format(java.lang.Double.valueOf(u!!.avaliacao)).toString()
                if (u!!.urlPerfil != null) {
                    Glide.with(applicationContext).load(u!!.urlPerfil).into(imgPerfil)
                }
                if (u!!.avaliacao.equals("0")) {
                    imgStar1.setBackgroundResource(R.drawable.starvoid)
                    imgStar2.setBackgroundResource(R.drawable.starvoid)
                    imgStar3.setBackgroundResource(R.drawable.starvoid)
                    imgStar4.setBackgroundResource(R.drawable.starvoid)
                    imgStar5.setBackgroundResource(R.drawable.starvoid)
                } else if (java.lang.Double.valueOf(u!!.avaliacao) < 2 && java.lang.Double.valueOf(u!!.avaliacao) >= 1) {
                    imgStar1.setBackgroundResource(R.drawable.staryellow)
                } else if (java.lang.Double.valueOf(u!!.avaliacao) < 3 && java.lang.Double.valueOf(u!!.avaliacao) >= 2) {
                    imgStar1.setBackgroundResource(R.drawable.staryellow)
                    imgStar2.setBackgroundResource(R.drawable.staryellow)
                } else if (java.lang.Double.valueOf(u!!.avaliacao) < 4 && java.lang.Double.valueOf(u!!.avaliacao) >= 3) {
                    imgStar1.setBackgroundResource(R.drawable.staryellow)
                    imgStar2.setBackgroundResource(R.drawable.staryellow)
                    imgStar3.setBackgroundResource(R.drawable.staryellow)
                } else if (java.lang.Double.valueOf(u!!.avaliacao) < 5 && java.lang.Double.valueOf(u!!.avaliacao) >= 4) {
                    imgStar1.setBackgroundResource(R.drawable.staryellow)
                    imgStar2.setBackgroundResource(R.drawable.staryellow)
                    imgStar3.setBackgroundResource(R.drawable.staryellow)
                    imgStar4.setBackgroundResource(R.drawable.staryellow)
                } else {
                    imgStar1.setBackgroundResource(R.drawable.staryellow)
                    imgStar2.setBackgroundResource(R.drawable.staryellow)
                    imgStar3.setBackgroundResource(R.drawable.staryellow)
                    imgStar4.setBackgroundResource(R.drawable.staryellow)
                    imgStar5.setBackgroundResource(R.drawable.staryellow)
                }
            }

            override fun onCancelled(databaseError: DatabaseError) {

            }
        })
    }

    private fun inicializarFirebase() {
        FirebaseApp.initializeApp(this@ProfActivity)
        bd = FirebaseDatabase.getInstance("https://imake-b81bb.firebaseio.com/")
        referenceBd = bd!!.reference
        referenceBd2 = bd!!.reference

        mStorageRef = FirebaseStorage.getInstance().getReference("uploads")
        mDatabaseRef = FirebaseDatabase.getInstance().getReference("Usuario")
    }
}
